### ISSUE TITLE
Documentation of XMany 

### ISSUE THREAD

**Issue by asgerhallas** [sprache/Sprache#46]
Small issue, I know... but the XMany method is documented as: "Parse a stream of elements. If any element is partially parsed". I have a feeling that something is missing?

And unfortunately I don't understand how it works. Could anyone help me with a less brief summary? :)

**Comment by ronnieoverby** [comment]
+1

I'm still trying to work out what the X* methods do.

**Comment by gordonwatts** [comment]
I was having this same question as well. So I did a little digging. In general, an upgrade of the xml comments would be welcome, but I suspect any work would be gladly accepted as a pull request by the authors... :-)

It boils down to first understanding how XOr works:

         /// <summary> 
         /// Parse first, if it succeeds, return first, otherwise try second. 
         /// Assumes that the first parsed character will determine the parser chosen (see Try). 
         /// </summary> 
         public static Parser<T> XOr<T>(this Parser<T> first, Parser<T> second) 

So that means if first accepts even a single character, then you are committed. If it fails, then the XOr will fail as well, even if the second parser would have succeeded (the Or parser would have handled that case).

I'm not 100% where you'd use these in parsing, actually. :-)

Then, the XMany code is actually pretty simple, so perhaps I can figure it out now:

         public static Parser<IEnumerable<T>> XMany<T>(this Parser<T> parser) 
         { 
             if (parser == null) throw new ArgumentNullException("parser"); 
  
             return parser.Many().Then(m => parser.Once().XOr(Return(m))); 
         } 

So it looks like it will parse as many items as it can, and if the character starting after the last successful matches the first one in the parser, it will fail.

I think that means if your parser looks for "ab" as Parse.String("ab").XMany(), then it will successfully match the "ab"'s in "abababababdude", but it will fail on the second "a" in the "aa" in "ababababaab".

In short, it seems like a way of terminating parsing if there is a partial match. But people who use parsers more than I should answer that.

Perhaps the doc should say "Parse a stream of elements. If any element is only partially parsed, fail"?

**Comment by nblumhardt** [comment]
Great writeup, thanks! The use case for these is in improving the locality of error messages; if it's possible to express a grammar using `X*` methods it'll generally produce better output/be easier to debug. They're not the default because the alternative behaviour is more general/forgiving. Doc and/or wiki improvements accepted, as you suggest :-) Cheers!

**Comment by asgerhallas** [comment]
Thanks for both your answers! And sorry @gordonwatts for the late reply, I have somehow overlooked yours. I will make a pull request for the missing documentation, then you can judge if I've got i right :)

**Comment by gordonwatts** [comment]
Sweet! And thanks! And thanks for the info on the use case. Actually, that is very nice - the biggest problem I have with Sprache is the error messages when a failure occurs. Usually I just get an error at the first character of the file, despite the error actually being a good part of the way in. :( Once day I'll learn!

**Comment by dueringa** [comment]
Would it make sense to include @gordonwatts example as an XML-doc `<example>`, to illustrate the differences?
